Add summary form logic (#1792)
workfloworchestrator/orchestrator-core: docs/reference-docs/workflows/summary-form.md

# Summary Forms

> 

Reviewer: Maybe add an adminition here to emphasize that this is new

Before a create or modify workflow submits its input, it is common to show the user a
read-only recap of everything that is about to change: the values just entered, and — for modify
workflows — what they looked like before. `orchestrator.core.forms.summary_form` provides a
small toolkit for building this recap page (a "summary form") without having to hand-write a
`FormPage` for every workflow.

The recap is rendered by the frontend as one or more read-only tables, using the same

> `MigrationSummary` field type that regular `FormPage`s use for any other input.

Reviewer: Do we really use migrationsummary for input?

## Quick start

For the common case — a single product table, optionally compared against its previous values —
use `base_summary`. It is a generator, so it must be delegated to with `yield from` at the end of
an `initial_input_form_generator`:

```python
from orchestrator.core.forms.summary_form import base_summary

def initial_input_form_generator(product: UUID, product_name: str) -> FormGenerator:
    class SelectOptionsForm(FormPage):
        model_config = ConfigDict(title=product_name)

        speed: int
        vlan: int

    user_input = yield SelectOptionsForm

    yield from base_summary(product_name, user_input.model_dump())

    return user_input.model_dump()

> ```

Reviewer: Please adapt all code blocks to have version tabs, like the rest of the documentation. In this case, the minimum version can be >= 5.2.0

This yields a `FormPage` titled `f"{product_name} Summary"` containing one table with a row per
field of `user_input`.

!!! note "Non-data fields in `new_data`/`old_data`"
    Passing a whole form's dump straight in (as above) also includes its non-data fields, e.g.
    `Divider` and `Label` fields. Any field name *starting with* `divider` (e.g. `divider_1`, the
    by-convention name) or `label` (e.g. `label_settings`) is dropped automatically — the check is
    a prefix match, not a substring match, so a real data field like `label_color` is excluded but
    `my_label` or `sub_divider` are not. Fields with a real value that need more context
    (e.g. showing `customer_id` as a customer name instead of a raw UUID) should use a formatter

>     instead, see below.

Reviewer: As discussed, remove/rewrite in favor of passing through `exclude` param

For a modify workflow, pass the subscription's current values as `old_data`
to render a **before / after** table instead:

```python

> yield from base_summary(product_name, new_data=user_input.model_dump(), old_data=previous_values)

Reviewer: Full example would be nice

| `single_column=True` | One single-column table per item, instead of one combined table |

Use `single_column=True` when the items don't share a natural "before/after" or "side by side"
relationship and read better as their own standalone tables (for example, a list of dissimilar
endpoints).

Per-item tables (before/after and single-column) are numbered by their position in `data` by
default: the first item is `"item_1"`, the second `"item_2"`, and so on. That default breaks down
once items carry an identity of their own that predates the summary — for example items 1, 2 and 3
of a subscription, where the user removes item 2. Renumbering by position would relabel the
remaining two as "item_1" and "item_2", even though the surviving item is still item 3 everywhere
else (error messages, the previous form pages, `old_data`). Set `TABLE_NUMBER_FIELD`
(`"__table_number"`) on the item's `new` dict to a number it should keep instead:

```python
tables = [
    TableOptions(
        name="item",
        data=make_table_data(
            [{**item, TABLE_NUMBER_FIELD: original_index} for original_index, item in surviving_items]
        ),
    ),
]
```

> `TABLE_NUMBER_FIELD` is excluded from the rendered rows like any other non-data field.

Reviewer: Imho this is too much implementation detail

|---|---|
| `name` | Table title, and field name prefix on the generated form |
| `data` | `Sequence[(new: dict, old: dict \| None)]` — the rows to render |
| `exclude` | Field names to omit from every row |
| `formatter` | Per-field `Formatter` overrides, see below |
| `header` | `Callable[[int], str]` to compute a column header from its 1-based index |
| `empty_message` | Message shown when `data` is empty |
| `single_column` | Render one table per item instead of one combined table |

`SummaryOptions` additionally accepts `after_product` (extra fields merged in after the product
table, e.g. to show a callout) and `tables` (the list of extra `TableOptions` described above).

## Custom field formatters

By default, a field is shown as its field name and `str(value)`; the frontend translates the field
name the same way it translates any other form field (falling back to the raw name if no
translation exists — `forms.fields.summary.<key>` is checked before the shared `forms.fields.<key>`
entry, see `en-GB.json`). Some fields need richer rendering — for example a subscription ID that
should show the linked subscription's description, or a composite value that should expand into

> several rows. A `Formatter` does this:

Reviewer: Give it a bit of a human touch

```suggestion
By default, a field is shown as its field name and `str(value)`; the frontend translates the field
name the same way it translates any other form field (falling back to the raw name if no
translation exists).

For fields that need custom rendering, for example a subscription ID that
shows the linked subscription's description or a composite value that expands into
several rows, a `Formatter` can be used:
```

```python
from collections.abc import Generator

RowGenerator = Generator[tuple[str, str], None, None]

def notification_summary(notification: dict) -> RowGenerator:

>     """Expand a single `notification` field into three summary rows."""

Reviewer: I count 2 rows?

    enabled = notification["enabled"]
    yield "Notifications enabled", str(enabled)
    yield "Channel", notification["channel"] if enabled else "N/A"
```

A `Formatter` is any `Callable[[Any], RowGenerator]`: given the field's value, it yields

> `(label, value)` pairs — usually one, but as many as needed.

Reviewer: ```suggestion
A `Formatter` is any `Callable[[Any], RowGenerator]` that given the field's value yields one or more `(label, value)` pairs.
```

> There are two ways to plug a formatter in:

Reviewer: ```suggestion
There are two ways to use a formatter:
```

* **Per table**, via `TableOptions(formatter={"notification": notification_summary, ...})` — only
  applies to that table.
* **Globally**, by adding it to `DEFAULT_FORMATTERS` — applies to every summary table in the

>   process, keyed by field name:

Reviewer: ```suggestion
* **Globally**, by adding it to `DEFAULT_FORMATTERS` — applies to every summary table in your workflows, keyed by field name:
```

  ```

  With `sap_summary` registered, a table of endpoints where each item has a `sap` field renders the
  port's `subscription_id`, `description` and `title` plus `vlan` for every endpoint, instead of a
  single row with the raw `sap` dict:

  ```python
  tables = [
      TableOptions(
          name="endpoint",
          data=make_table_data([{"sap": endpoint.sap} for endpoint in new_endpoints]),
      ),
  ]
  ```

## API summary

All of the above is importable from `orchestrator.core.forms.summary_form`:

| Name | Purpose |
|---|---|
| `base_summary` | Single product table, with optional before/after comparison |
| `generate_summary_form` | Full control via `SummaryOptions` / `TableOptions` |
| `create_table` | Build a single table's `MigrationSummary` field directly |
| `make_table_data` | Zip new/old item lists into `TableOptions["data"]` |
| `SummaryOptions`, `TableOptions`, `BaseOptions` | Option `TypedDict`s |
| `TABLE_NUMBER_FIELD` | Item dict key that overrides a per-item table's displayed number |
| `DEFAULT_FORMATTERS` | Global, mutable field-name → `Formatter` registry |
| `Formatter`, `RowGenerator` | Type aliases for writing your own formatters |

> | `customer_name_summary_field`, `select_list_summary`, `subscription_summary_fields` | Built-in formatter helpers |

Reviewer: Can we let mkdocs parse this directly from the `orchestrator.core.forms.summary_form` file? Like we're doing for the CLI docs
